pkg/plugin/plugin.go

```diff
@@ -447,57 +473,77 @@ func cmdAdd(args *skel.CmdArgs) error {
 	}
 
 	// Print result to stdout, in the format defined by the requested cniVersion.
-	return cnitypes.PrintResult(result, conf.CNIVersion)
+	err = cnitypes.PrintResult(result, conf.CNIVersion)
+	return
 }
 
-func cmdDel(args *skel.CmdArgs) error {
+func cmdDel(args *skel.CmdArgs) (err error) {
+	// Defer a panic recover, so that in case we panic we can still return
+	// a proper error to the runtime.
+	defer func() {
+		if e := recover(); e != nil {
+			msg := fmt.Sprintf("Calico CNI panicked during DEL: %s", e)
+			if err != nil {
+				// If we're recovering and there was also an error, then we need to
+				// present both.
+				msg = fmt.Sprintf("%s: error=%s", msg, err)
+			}
+			err = fmt.Errorf(msg)
+		}
+	}()
+
 	conf := types.NetConf{}
-	if err := json.Unmarshal(args.StdinData, &conf); err != nil {
-		return fmt.Errorf("failed to load netconf: %v", err)
+	if err = json.Unmarshal(args.StdinData, &conf); err != nil {
+		err = fmt.Errorf("failed to load netconf: %v", err)
+		return
 	}
 
 	utils.ConfigureLogging(conf.LogLevel)
 
 	if !conf.NodenameFileOptional {
 		// Configured to wait for the nodename file - don't start until it exists.
-		if _, err := os.Stat("/var/lib/calico/nodename"); err != nil {
+		if _, err = os.Stat("/var/lib/calico/nodename"); err != nil {
 			s := "%s: check that the calico/node container is running and has mounted /var/lib/calico/"
-			return fmt.Errorf(s, err)
+			err = fmt.Errorf(s, err)
+			return
 		}
 		logrus.Debug("/var/lib/calico/nodename exists")
 	}
 
 	// Determine which node name to use.
 	nodename := utils.DetermineNodename(conf)
 
-	epIDs, err := utils.GetIdentifiers(args, nodename)
+	var epIDs *utils.WEPIdentifiers
+	epIDs, err = utils.GetIdentifiers(args, nodename)
```

Comment:
Do we need to change the "short declaration" syntax here? I thought that Go would redeclare existing vars (`err`) in this case and not overwrite them but I'm not sure.

Reply:
I guess the long form is more obviously correct; in that case, maybe we can do that same in cmdAdd too